contracts/starknet/space/space.cairo

```diff
@@ -0,0 +1,214 @@
+func proposal_threshold() -> (threshold : Uint256):
+end
+
+# TODO: Should be Address not felt
+@storage_var
+func voting_strategy() -> (strategy_address : felt):
+end
+
+@storage_var
+func authenticator() -> (authenticator_address : felt):
+end
+
+@storage_var
+func next_proposal_nonce() -> (nonce : felt):
+end
+
+@storage_var
+func proposal_registry(proposal_id : felt) -> (proposal : Proposal):
+end
+
+@storage_var
+func vote_registry(proposal_id : felt, voter_address : EthAddress) -> (vote : Vote):
+end
+
+@storage_var
+func vote_power(proposal_id : felt, choice : felt) -> (power : Uint256):
+end
+
+@event
+func proposal_created(proposal_id : felt, proposer_address : EthAddress, proposal : Proposal):
+end
+
+@event
+func vote_created(proposal_id : felt, voter_address : EthAddress, vote : Vote):
+end
+
+# Throws if the caller address is not identical to the authenticator address (stored in the `authenticator` variable)
+func authenticator_only{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}():
+    let (caller_address) = get_caller_address()
+    let (authenticator_address) = authenticator.read()
+
+    # Ensure it has been initialized
+    assert_not_zero(authenticator_address)
+    # Ensure the caller is the authenticator contract
+    assert caller_address = authenticator_address
+
+    return ()
+end
+
+@constructor
+func constructor{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr : felt}(
+        _voting_delay : felt, _voting_period : felt, _proposal_threshold : Uint256,
+        _voting_strategy : felt, _authenticator : felt):
+    # Sanity checks
+    assert_nn(_voting_delay)
+    assert_nn(_voting_period)
+    assert_not_zero(_voting_strategy)
+    assert_not_zero(_authenticator)
+    # TODO: maybe use uint256_signed_nn to check proposal_threshold?
+
+    # Initialize the storage variables
+    voting_delay.write(_voting_delay)
+    voting_period.write(_voting_period)
+    proposal_threshold.write(_proposal_threshold)
+    voting_strategy.write(_voting_strategy)
+    authenticator.write(_authenticator)
+    next_proposal_nonce.write(1)
+
+    return ()
+end
+
+@external
+func vote{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr : felt}(
+        voter_address : EthAddress, proposal_id : felt, choice : felt) -> ():
+    alloc_locals
+
+    # Verify that the caller is the authenticator contract.
+    authenticator_only()
+
+    let (proposal) = proposal_registry.read(proposal_id)
+    let (current_timestamp) = get_block_timestamp()
+
+    # Make sure proposal is not closed
+    assert_lt(current_timestamp, proposal.end_timestamp)
+
+    # Make sure proposal has started
+    assert_le(proposal.start_timestamp, current_timestamp)
+
+    # Make sure voter has not already voted
+    let (prev_vote) = vote_registry.read(proposal_id, voter_address)
+    if prev_vote.choice != 0:
+        # Voter has already voted!
+        assert 1 = 0
+    end
+
+    let (local params : felt*) = alloc()
```

Comment:
We should pass these params as an argument for `vote`, so that the interface doesnt change once parameters are needed. Will just be an empty array for milestone 1.

Reply:
[Use params in execute, vote, and proposal](https://github.com/snapshot-labs/sx-core/pull/47/commits/a3a55bd898bbf1061ecd8e15041e4ca7a6a6ed80) :D